<div>
<img src="header5.png" width="1000" float='left', style="display:block">
</div>
<a id='tartalomjegyzek'></a>

# Üzleti Intelligencia - Adatelemzési Feladat

**Radó László, MS1PB5. 2020 Április**
<br>
<hr>
<hr>

## 3) Modellezés

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb
import matplotlib.patches as mpatches
import matplotlib.ticker as mtick

%matplotlib inline

no_show =  pd.read_csv('KaggleV2-May-2016_clean.csv', 
                       index_col='AppointmentID',
                       infer_datetime_format=True, 
                       parse_dates=['ScheduledDay', 'AppointmentDay'])

In [230]:
keep_cols = [
#  'Unnamed: 0',
#  'PatientId',
#  'Gender',
#  'ScheduledDay',
#  'AppointmentDay',
 'Age',
#  'Neighborhood',
 'Osztondij',
 'Magas_vernyomas',
 'Cukorbeteseg',
 'Alkoholizmus',
 'Fogyatekossag',
 'SMS_ertesito',
#  'No-show',
#  'dummy_No-show',
 'dummy_Gender',
 'Fogyatekossag_binary',
 'betegsegek_szama',
 'dummy_AEROPORTO',
 'dummy_ANDORINHAS',
 'dummy_ANTÔNIO_HONÓRIO',
 'dummy_ARIOVALDO_FAVALESSA',
 'dummy_BARRO_VERMELHO',
 'dummy_BELA_VISTA',
 'dummy_BENTO_FERREIRA',
 'dummy_BOA_VISTA',
 'dummy_BONFIM',
 'dummy_CARATOÍRA',
 'dummy_CENTRO',
 'dummy_COMDUSA',
 'dummy_CONQUISTA',
 'dummy_CONSOLAÇÃO',
 'dummy_CRUZAMENTO',
 'dummy_DA_PENHA',
 'dummy_DE_LOURDES',
 'dummy_DO_CABRAL',
 'dummy_DO_MOSCOSO',
 'dummy_DO_QUADRO',
 'dummy_ENSEADA_DO_SUÁ',
 'dummy_ESTRELINHA',
 'dummy_FONTE_GRANDE',
 'dummy_FORTE_SÃO_JOÃO',
 'dummy_FRADINHOS',
 'dummy_GOIABEIRAS',
 'dummy_GRANDE_VITÓRIA',
 'dummy_GURIGICA',
 'dummy_HORTO',
 'dummy_ILHA_DAS_CAIEIRAS',
 'dummy_ILHA_DE_SANTA_MARIA',
 'dummy_ILHA_DO_BOI',
 'dummy_ILHA_DO_FRADE',
 'dummy_ILHA_DO_PRÍNCIPE',
 'dummy_ILHAS_OCEÂNICAS_DE_TRINDADE',
 'dummy_INHANGUETÁ',
 'dummy_ITARARÉ',
 'dummy_JABOUR',
 'dummy_JARDIM_CAMBURI',
 'dummy_JARDIM_DA_PENHA',
 'dummy_JESUS_DE_NAZARETH',
 'dummy_JOANA_D´ARC',
 'dummy_JUCUTUQUARA',
 'dummy_MARIA_ORTIZ',
 'dummy_MARUÍPE',
 'dummy_MATA_DA_PRAIA',
 'dummy_MONTE_BELO',
 'dummy_MORADA_DE_CAMBURI',
 'dummy_MÁRIO_CYPRESTE',
 'dummy_NAZARETH',
 'dummy_NOVA_PALESTINA',
 'dummy_PARQUE_INDUSTRIAL',
 'dummy_PARQUE_MOSCOSO',
 'dummy_PIEDADE',
 'dummy_PONTAL_DE_CAMBURI',
 'dummy_PRAIA_DO_CANTO',
 'dummy_PRAIA_DO_SUÁ',
 'dummy_REDENÇÃO',
 'dummy_REPÚBLICA',
 'dummy_RESISTÊNCIA',
 'dummy_ROMÃO',
 'dummy_SANTA_CECÍLIA',
 'dummy_SANTA_CLARA',
 'dummy_SANTA_HELENA',
 'dummy_SANTA_LUÍZA',
 'dummy_SANTA_LÚCIA',
 'dummy_SANTA_MARTHA',
 'dummy_SANTA_TEREZA',
 'dummy_SANTO_ANDRÉ',
 'dummy_SANTO_ANTÔNIO',
 'dummy_SANTOS_DUMONT',
 'dummy_SANTOS_REIS',
 'dummy_SEGURANÇA_DO_LAR',
 'dummy_SOLON_BORGES',
 'dummy_SÃO_BENEDITO',
 'dummy_SÃO_CRISTÓVÃO',
 'dummy_SÃO_JOSÉ',
 'dummy_SÃO_PEDRO',
 'dummy_TABUAZEIRO',
 'dummy_UNIVERSITÁRIO',
 'dummy_VILA_RUBIM',
 'ScheduledHour',
 'DayofWeek',
 'Num_days_to_appointment',
 'number_of_visits',
 'frequent_visitor',
 'number_of_shows',
 'frequent_show',
 'number_of_no_shows',
 'frequent_no_show',
 'prop_show',
#  'prop_no_show',
 'prop_show_noshow',
 'Age_capped_80',
 'betegsegek_szama_capped',
 'number_of_no_shows_capped',
 'number_of_shows_capped',
 'number_of_visits_capped',
 'number_of_visits_capped_mult',
 'Num_days_to_appointment_capped',
 'dummy_hour__6',
 'dummy_hour__7',
 'dummy_hour__8',
 'dummy_hour__9',
 'dummy_hour__10',
 'dummy_hour__11',
 'dummy_hour__12',
 'dummy_hour__13',
 'dummy_hour__14',
 'dummy_hour__15',
 'dummy_hour__16',
 'dummy_hour__17',
 'dummy_hour__18',
 'dummy_hour__19',
 'dummy_hour__20',
 'dummy_hour__21',
#  'Age_Quantile',
 'number_of_shows_capped_sq',
 'number_of_shows_sq',
 'number_of_visits_capped_sq',
 'number_of_visits_sq',
 'number_of_visits_capped_mult_sq',
 'prop_show_log',
#  'prop_no_show_log',
 'prop_show_noshow_log'
]

X = no_show[keep_cols]
y = no_show['dummy_No-show']

In [543]:
def __evaluate(parameters):
    
    key = parameters['key']
    y_test = parameters['y_test']
    y_pred = parameters['y_pred']
    pca = parameters['pca']
    scale = parameters['scale']
    regul = parameters['regul']
    columns = parameters['results']
    
    # Confusion Mátrix értékeinek eltárolása változókban.
    from sklearn.metrics import confusion_matrix, accuracy_score, recall_score, precision_score
    
    cm = confusion_matrix(y_test, y_pred)
    fp = cm[0][1]
    fn = cm[1][0]
    tp = cm[1][1]
    tn = cm[0][0]
    
    # Értékek hozzárendelése a kimeneti táblához.
    columns['Model'].append(key)
    columns['Accuracy'].append(np.round(accuracy_score(y_test, y_pred),4))
    
    # Specificity és Recall
    spec = np.round(cm[0][0] / (cm[0][0] + cm[0][1]),4)
    columns['Specificity'].append(spec)
    columns['Recall'].append(np.round(recall_score(y_test, y_pred),4))
    
    # Confusion Mátrix értékei
    columns['True Positive'].append(tp)
    columns['False Positive'].append(fp)
    columns['True Negative'].append(tn)
    columns['False Negative'].append(fn)
    
    # Adatelőkészítési beállítások
    columns['PCA'].append(pca)
    columns['Scaling'].append(scale)
    columns['Regul.'].append(regul)
        
    return columns

In [544]:
def iterate_models(X, y, classifiers, pca=False, scale=False, regul=False, dec_bound=0.5):
    import time
  
    if regul == True:
        from sklearn.feature_selection import SelectFromModel
        sel_ = SelectFromModel(LogiReg(max_iter=5000, random_state = 42))
        sel_.fit(X, y)
        selected_features = X.columns[(sel_.get_support())]
        X = X[selected_features]

        
    if scale == True:
        from sklearn.preprocessing import StandardScaler
        sc = StandardScaler()
        X = sc.fit_transform(X)
        
    
    from sklearn.model_selection import train_test_split
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

    
    if pca != False:
        from sklearn.decomposition import PCA
        pc_analyser = PCA(n_components=pca, random_state = 42)
        X_train = pc_analyser.fit_transform(X_train)
        X_test = pc_analyser.transform(X_test)

        
    if type(dec_bound) == float:
        start = int(dec_bound * 100)
        stop = start + 1
        step = 1
    else:
        start = int(dec_bound[0] * 100)
        stop = int(dec_bound[1] * 100)
        step = int(dec_bound[2] * 100)
        
    y_preds = []
    
    parameters = {
                 'key' : '',
                 'y_test' : y_test,
                 'y_pred' : '',
                 'dec_bound' : dec_bound,
                 'pca' : pca,
                 'scale' : scale,
                 'regul' : regul,
                 'results' : {
                     'Model' : [],
                     'Accuracy' : [],
                     'Specificity': [],
                     'Recall' : [],
                     'Decision Boundary' : [],
                     'True Positive' : [], 
                     'False Positive' : [], 
                     'True Negative' : [], 
                     'False Negative' : [],
                     'PCA' : [],
                     'Scaling' : [],
                     'Regul.' : [],
                     'Runtime' : [],
                     'Parameteres' : []
                    }
    }
       
    
    for key in classifiers:
        start_time = time.time()
        
        parameters['key'] = key
        params = str(classifiers[key]).replace(str(key), '')

        classifier = classifiers[key]
        classifier.fit(X_train, y_train)
                
        if key == 'LogisticRegression' or key == 'GaussianNB':
            for i in range(start,stop,step):
                
                dec__bound_iterate = i/100
                parameters['results']['Decision Boundary'].append(dec__bound_iterate)

                y_pred_prob = classifier.predict_proba(X_test)
                y_pred = [1 if prob[1] >= dec__bound_iterate else 0 for prob in y_pred_prob]
                y_preds.append(y_pred)
                parameters['y_pred'] = y_pred
                
                parameters['results'] = __evaluate(parameters)
                parameters['results']['Runtime'].append(time.time() - start_time)
                parameters['results']['Parameteres'].append(params)
                
        else:
            y_pred = classifier.predict(X_test)
            y_preds.append(y_pred)
            parameters['y_pred'] = y_pred
            parameters['results'] = __evaluate(parameters)
            parameters['results']['Decision Boundary'].append('n/a')
            parameters['results']['Runtime'].append(time.time() - start_time)
            parameters['results']['Parameteres'].append(params)
            
    return pd.DataFrame.from_dict(parameters['results']), y_preds

In [45]:
from sklearn.linear_model import LogisticRegression as LogiReg
from sklearn.neighbors import KNeighborsClassifier as KNC
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB as Naive
from sklearn.ensemble import RandomForestClassifier as RFC
from sklearn.ensemble import GradientBoostingClassifier as GBC
from xgboost import XGBClassifier as XGBC

In [545]:
classifiers = {
               'LogisticRegression': LogiReg(max_iter=1000),
#                'KNeighborsClassifier': KNC(n_neighbors=8, leaf_size=10, n_jobs=-1),
#                'SVC': SVC(C=10.0, kernel='sigmoid', degree=10, gamma='scale', random_state=42),
#                'GaussianNB': Naive(),
#                'RandomForestClassifier': RFC(n_estimators=10, random_state=42),
#                'GradientBoostingClassifier': GBC(learning_rate=0.1, n_estimators=1000, random_state=42),
#                'XGBClassifier': XGBC(use_label_encoder=False, eval_metric='logloss')
              }

results, y_preds = iterate_models(X[:10000], y[:10000], classifiers, pca=27, scale=True, regul=True, dec_bound=0.5)
# results = pd.concat([results, results2])
results.reset_index()

index                       Model  Accuracy  Specificity  Recall  \
0      0          LogisticRegression    0.7910       0.0000  1.0000   
1      1        KNeighborsClassifier    0.7505       0.1603  0.9064   
2      2                         SVC    0.6920       0.1603  0.8325   
3      3                  GaussianNB    0.7605       0.0885  0.9381   
4      4      RandomForestClassifier    0.7875       0.0478  0.9829   
5      5  GradientBoostingClassifier    0.7820       0.0359  0.9791   
6      6               XGBClassifier    0.7900       0.0263  0.9918   

  Decision Boundary  True Positive  False Positive  True Negative  \
0               0.5           1582             418              0   
1               n/a           1434             351             67   
2               n/a           1317             351             67   
3               0.5           1484             381             37   
4               n/a           1555             398             20   
5               n/a           1549             403             15   
6               n/a           1569             407             11   

   False Negative  PCA  Scaling  Regul.    Runtime  \
0               0   27     True    True   0.057004   
1             148   27     True    True   1.136988   
2             265   27     True    True   3.402030   
3              98   27     True    True   0.017961   
4              27   27     True    True   0.181995   
5              33   27     True    True  34.502046   
6              13   27     True    True   1.141007   

                                         Parameteres  
0                                    (max_iter=1000)  
1           (leaf_size=10, n_jobs=-1, n_neighbors=8)  
2  (C=10.0, degree=10, kernel='sigmoid', random_s...  
3                                                 ()  
4                 (n_estimators=10, random_state=42)  
5               (n_estimators=1000, random_state=42)  
6  (base_score=None, booster=None, colsample_byle...

### Model eredmények 1:
* Látható, hogy a modellek nagyrésze megközelíti a 80%-os accuracy értéket.
* Viszont a specificitás (Valós Negatív Ráta), nagyon alacsony.
* Ennek oka a következő:
   * A modellünkben kiegyensúlyozatnal a célváltozó pozitív / negatív aránya (20% negatív).
   * Ezen kívül a független változóink magyarázó ereje alacsony.
   * Ezért a modellek stratégiája, hogy közel minden megfigyelést pozitívnak értékelnek.


* **Lehetséges megoldás:**
   * Probabilisztikus modellek döntési határának módosítása
   * **Cél:** Specificity javítása, miközben az accuracy értéket továbbra is magasan tartjuk
   
<hr>
Specificity = True Negatives / (True Negatives + False Positives)

In [629]:
classifiers = {
               'LogisticRegression': LogiReg(max_iter=1000, C=10.0, random_state=42),
#                'GaussianNB': Naive()
              }

#Gauss
# results_prob_2, y_preds = iterate_models(X, y, classifiers, pca=37, scale=True, regul=True, dec_bound=0.95)

#Logi
results_prob_2, y_preds = iterate_models(X, y, classifiers, pca=37, scale=True, regul=False, dec_bound=0.88)
results_prob = pd.concat([results_prob, results_prob_2])
results_prob.reset_index()

index               Model  Accuracy  Specificity  Recall  \
0       0  LogisticRegression    0.7829       0.1030  0.9558   
1       0  LogisticRegression    0.7067       0.4831  0.7636   
2       0  LogisticRegression    0.5528       0.7977  0.4904   
3       0  LogisticRegression    0.2095       0.9996  0.0084   
4       0          GaussianNB    0.7240       0.2962  0.8329   
5       0          GaussianNB    0.6394       0.4978  0.6754   
6       0          GaussianNB    0.5372       0.7281  0.4886   
7       0          GaussianNB    0.2769       0.9367  0.1090   
8       0          GaussianNB    0.7171       0.3363  0.8140   
9       0  LogisticRegression    0.6458       0.6416  0.6468   
10      0  LogisticRegression    0.5778       0.7683  0.5293   
11      0  LogisticRegression    0.5212       0.8260  0.4436   
12      0  LogisticRegression    0.4785       0.8595  0.3816   
13      0  LogisticRegression    0.4282       0.8927  0.3100   
14      0  LogisticRegression    0.4785       0.8595  0.3816   
15      0  LogisticRegression    0.3723       0.9237  0.2319   
16      0  LogisticRegression    0.4282       0.8927  0.3100   

    Decision Boundary  True Positive  False Positive  True Negative  \
0                0.60          16843            4022            462   
1                0.75          13456            2318           2166   
2                0.85           8642             907           3577   
3                0.94            148               2           4482   
4                0.60          14676            3156           1328   
5                0.75          11901            2252           2232   
6                0.85           8610            1219           3265   
7                0.95           1920             284           4200   
8                0.80          14344            2976           1508   
9                0.80          11398            1607           2877   
10               0.84           9327            1039           3445   
11               0.86           7817             780           3704   
12               0.87           6724             630           3854   
13               0.88           5462             481           4003   
14               0.87           6724             630           3854   
15               0.89           4087             342           4142   
16               0.88           5462             481           4003   

    False Negative  PCA  Scaling  Regul.   Runtime  \
0              778   37     True   False  0.252003   
1             4165   37     True   False  0.260001   
2             8979   37     True   False  0.290002   
3            17473   37     True   False  0.273998   
4             2945   37     True    True  0.264048   
5             5720   37     True    True  0.430006   
6             9011   37     True    True  0.250182   
7            15701   37     True    True  0.390005   
8             3277   37     True   False  0.263964   
9             6223   37     True   False  0.336115   
10            8294   37     True   False  0.268002   
11            9804   37     True   False  0.305008   
12           10897   37     True   False  0.259001   
13           12159   37     True   False  0.242006   
14           10897   37     True   False  0.388994   
15           13534   37     True   False  0.323006   
16           12159   37     True   False  0.257001   

                                 Parameteres  
0   (C=10.0, max_iter=1000, random_state=42)  
1   (C=10.0, max_iter=1000, random_state=42)  
2   (C=10.0, max_iter=1000, random_state=42)  
3   (C=10.0, max_iter=1000, random_state=42)  
4                                         ()  
5                                         ()  
6                                         ()  
7                                         ()  
8                                         ()  
9   (C=10.0, max_iter=1000, random_state=42)  
10  (C=10.0, max_iter=1000, random_state=42)  
11  (C=10.0, max_iter=1000, random_

### Model eredménye 2:
* Látható, hogy a döntési határok eltolásával mindkét probabilisztikus model esetében jelentősen sikerült növelni a specificitást.
* Viszont a növekedés a Recall (érzékenység, Valós Pozitív Ráta) rovására ment.
* Ezért az accuracy érték is jelentősen csökkent.
* Az ideális döntési határ megállapítására üzleti tervet és költség / megtakarítás elemzést kell végezni.


In [636]:
model_df = results_prob
model_index = 16

tp = model_df['True Positive'].iloc[model_index] / 12
fp = model_df['False Positive'].iloc[model_index] / 12
tn = model_df['True Negative'].iloc[model_index] / 12
fn = model_df['False Negative'].iloc[model_index] / 12
total = tp + fp + tn + fn
total_n = tn + fn
spec = model_df['Specificity'].iloc[model_index]

havi_atlag_paciens = total
figyelmeztetes_arany = np.round(total_n / total,3)
figyelmeztetes_szama =  total_n

print(f'Figyelmeztetések aránya: {np.round(figyelmeztetes_arany,3)}')
print(f'Figyelmeztetések száma (havi): {int(figyelmeztetes_szama)}')
print('')
print(f'Elért potenciális meg nem jelenők aránya: {np.round(spec,3)}')
print(f'Elért potenciális meg nem jelenők száma (havi): {int(tn)}')
print(f'El nem ért potenciális meg nem jelenők száma (havi): {int(fp)}')
print(f'Potenciális meg nem jelenők száma (havi): {int(fp+tn)}')
print('')
print(f'Fölösleges figyelmeztetések aránya: {np.round((figyelmeztetes_szama-tn) / figyelmeztetes_szama,3)}')
print(f'Fölösleges figyelmeztetések száma (havi): {int(figyelmeztetes_szama-tn)}')
print(f'Sikeres figyelemeztetesek aránya: {np.round(tn / figyelmeztetes_szama,3)}')

Figyelmeztetések aránya: 0.731
Figyelmeztetések száma (havi): 1346

Elért potenciális meg nem jelenők aránya: 0.893
Elért potenciális meg nem jelenők száma (havi): 333
El nem ért potenciális meg nem jelenők száma (havi): 40
Potenciális meg nem jelenők száma (havi): 373

Fölösleges figyelmeztetések aránya: 0.752
Fölösleges figyelmeztetések száma (havi): 1013
Sikeres figyelemeztetesek aránya: 0.248


In [637]:
# költségek 
meg_nem_jelenes_ktg = 200
meg_nem_jelenes_okozta_ktg = meg_nem_jelenes_ktg * (tn+fp)

# Figyelmeztetes koltsegei
havi_hivasok_szama_per_alkalmazott = 2000
egy_hivas_ktg = 7
alkalmazott_ktg = 3000
model_uzem_ktg = 1500
egyeb_fix_ktg = 500

szukseges_alkalmazott = np.ceil(figyelmeztetes_szama / havi_hivasok_szama)

havi_ktg = szukseges_alkalmazott * (alkalmazott_ktg + egyeb_fix_ktg) \
         + model_uzem_ktg \
         + figyelmeztetes_szama * egy_hivas_ktg

# Azon potenciálisan meg nem jelenők aránya, akik a figyelmeztetés hatására mégis eljönnek
megjelenes_javulas = 0.3
javulas_szama = tn * megjelenes_javulas

megtakaritas_havi = int(javulas_szama * meg_nem_jelenes_ktg) - int(havi_ktg)

print(f'Össz uzemeltetési költség (havi): {int(havi_ktg)} USD')
print(f'Meg nem jelenések okozta költség (havi): {int(meg_nem_jelenes_okozta_ktg)} USD')
print('')
print(f'Ktg csökkenés, figyelmeztetés hatására (havi): {int(javulas_szama * meg_nem_jelenes_ktg)} USD')
print(f'Megtakarítás (havi): {megtakaritas_havi} USD')
print(f'Veszteségek csökkentése: {np.round(megtakaritas_havi / meg_nem_jelenes_okozta_ktg*100,1)}%')

Össz uzemeltetési költség (havi): 14427 USD
Meg nem jelenések okozta költség (havi): 74733 USD

Ktg csökkenés, figyelmeztetés hatására (havi): 20014 USD
Megtakarítás (havi): 5587 USD
Veszteségek csökkentése: 7.5%


In [625]:
szukseges_alkalmazott = np.ceil(total / havi_hivasok_szama)

havi_ktg = szukseges_alkalmazott * (alkalmazott_ktg + egyeb_fix_ktg) \
         + model_uzem_ktg \
         + total * egy_hivas_ktg

# Azon potenciálisan meg nem jelenők aránya, akik a figyelmeztetés hatására mégis eljönnek
javulas_szama = (tn+fp) * megjelenes_javulas

megtakaritas_havi = int(javulas_szama * meg_nem_jelenes_ktg) - int(havi_ktg)

print('EREDMÉNY MODELL HASZNÁLATA NÉLKÜL, HA MINDEN ÜGYFELET FELHÍVNÁNK:')
print('')
print(f'Össz uzemeltetési költség (havi): {int(havi_ktg)} USD')
print(f'Meg nem jelenések okozta költség (havi): {int(meg_nem_jelenes_okozta_ktg)} USD')
print('')
print(f'Ktg csökkenés, figyelmeztetés hatására (havi): {int(javulas_szama * meg_nem_jelenes_ktg)} USD')
print(f'Megtakarítás (havi): {megtakaritas_havi} USD')
print(f'Veszteségek csökkentése: {np.round(megtakaritas_havi / meg_nem_jelenes_okozta_ktg*100,1)}%')

EREDMÉNY MODELL HASZNÁLATA NÉLKÜL, HA MINDEN ÜGYFELET FELHÍVNÁNK:

Össz uzemeltetési költség (havi): 17894 USD
Meg nem jelenések okozta költség (havi): 74733 USD

Ktg csökkenés, figyelmeztetés hatására (havi): 22420 USD
Megtakarítás (havi): 4526 USD
Veszteségek csökkentése: 6.1%


### Eredmények értékelése:
* Ha minden ügyfelet felhívnánk, és így az összes potenciálisan meg nem jelenő pácienst elérnénk
* Valamint, ha a potenciálisan meg nem jelenő páciensek az emlékeztetés hatására 30%-kkal gyakrabban jönnek el (illetve más pácienseket tudunk az időpontjukra beosztani):
    * 6.1%-kkal tudnánk csökkenteni a meg nem jelenések okozta veszteségeket.
    * Ez 4552 USD megtakarítást jelentene havonta.
    

* **A modell használatával ezt az értéket sikerült a következők szerint javítani:**
    * 7.5%-kkal tudnánk csökkenteni a meg nem jelenések okozta veszteségeket.
    * Ez 5587 USD megtakarítást jelentene havonta.


* Ez körülbelül 1.4%P, azaz havonta körülbelül 1000 USD többlet megtakarítás

### Lehetséges további lépések:
* Mivel a független változóink magyarázó ereje nagyon alacsony, ezért nem sikerült hatékony modellt építeni a problémára.
* A modell további javítása érdekében a következő lépéseket tehetjük:
    * **Elsősorban a magyarázó változók bővítése, például:**
        * Milyen egészségügyi problémával kapcsolatos az előjegyzés (szívműtétre érkezik a páciens, vagy csak egy receptért jön)
        * Melyik egészségügyi osztályra szól az előjegyzés (sebészet, szemészet stb)
        * Az egészségügyi probléma súlyosságának osztályozása (1-3 skálán)
        * Kontroll, rendszeres, vagy első vizsgálat?
    * **További modellek építése:**
        * Neurális hálók
        * Halmozott ensemble model 
            * Különösen a probabilisztikus modellek valószínűség becslésének felhasználásával
            * Különböző döntési határokkal      
            
    * **További adatátalakítások:**
        * Interkaciók
        * Dummyzás